File: operator-dashboard/src/api/costs.ts
/**
 * Cost API Client
 *
 * Provides access to token usage and cost reporting endpoints.
 */
import apiClient from './client';

export interface ProjectCostSummary {
  projectId: string;
  projectName: string;
  totalRuns: number;
  totalPosts: number;
  totalInputTokens: number;
  totalOutputTokens: number;
  totalCacheCreationTokens: number;
  totalCacheReadTokens: number;
  totalGenerationCostUsd: number;
  totalResearchTools: number;
  totalResearchCostUsd: number;
  totalCostUsd: number;
  costPerPost: number | null;
}

export interface RunCostBreakdown {
  runId: string;
  projectId: string;
  status: string;
  startedAt: string;
  completedAt: string | null;
  totalInputTokens: number;
  totalOutputTokens: number;
  totalCacheCreationTokens: number;
  totalCacheReadTokens: number;
  totalCostUsd: number;
  estimatedCostUsd: number | null;
  totalPosts: number;
  postsWithTokenData: number;
  avgCostPerPost: number | null;
  cacheSavingsUsd: number | null;
}

export interface CostTrend {
  date: string;
  costUsd: number;
}

export interface UserCostSummary {
  userId: string;
  periodDays: number;
  totalProjects: number;
  totalRuns: number;
  totalInputTokens: number;
  totalOutputTokens: number;
  totalGenerationCostUsd: number;
  totalResearchTools: number;
  totalResearchCostUsd: number;
  totalCostUsd: number;
  topProjects: Array<{
    projectId: string;
    projectName: string;
    costUsd: number;
  }>;
  costTrend: CostTrend[];
}

export interface ResearchCostSummary {
  clientId: string;
  clientName: string;
  totalResearchTools: number;
  totalBusinessPriceUsd: number;
  totalActualCostUsd: number;
  priceDifferenceUsd: number;
  toolsBreakdown: Array<{
    toolName: string;
    toolLabel: string;
    executionCount: number;
    totalBusinessPrice: number;
    totalActualCost: number;
    totalTokens: number;
  }>;
}

export const costsApi = {
  /**
   * Get cost summary for a project
   */
  async getProjectCosts(projectId: string): Promise<ProjectCostSummary> {
    const response = await apiClient.get(`/api/costs/project/${projectId}`);
    return response.data;
  },

  /**
   * Get detailed cost breakdown for a run
   */
  async getRunCosts(runId: string): Promise<RunCostBreakdown> {
    const response = await apiClient.get(`/api/costs/run/${runId}`);
    return response.data;
  },

  /**
   * Get user-wide cost summary with trends
   */
  async getUserCostSummary(days: number = 30): Promise<UserCostSummary> {
    const response = await apiClient.get('/api/costs/summary', {
      params: { days },
    });
    return response.data;
  },

  /**
   * Get research cost summary for a client
   */
  async getResearchCosts(clientId: string): Promise<ResearchCostSummary> {
    const response = await apiClient.get(`/api/costs/research/${clientId}`);
    return response.data;
  },

  /**
   * Format token count with commas
   */
  formatTokens(tokens: number): string {
    return tokens.toLocaleString();
  },

  /**
   * Format USD cost with $ and 4 decimal places
   */
  formatCost(costUsd: number): string {
    return `$${costUsd.toFixed(4)}`;
  },

  /**
   * Format USD cost for display (2 decimals for larger amounts)
   */
  formatCostDisplay(costUsd: number): string {
    if (costUsd >= 1) {
      return `$${costUsd.toFixed(2)}`;
    }
    return `$${costUsd.toFixed(4)}`;
  },

  /**
   * Calculate cache savings percentage
   */
  calculateCacheSavingsPercent(
    cacheReadTokens: number,
    totalInputTokens: number
  ): number {
    if (totalInputTokens === 0) return 0;
    return (cacheReadTokens / totalInputTokens) * 100;
  },
};

File: backend/schemas/costs.py
"""
Cost API Response Schemas

Pydantic schemas for token usage and cost reporting endpoints.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ProjectCostSummary(BaseModel):
    """Cost summary for a single project"""

    project_id: str = Field(..., serialization_alias="projectId")
    project_name: str = Field(..., serialization_alias="projectName")
    total_runs: int = Field(..., serialization_alias="totalRuns")
    total_posts: int = Field(..., serialization_alias="totalPosts")
    total_input_tokens: int = Field(..., serialization_alias="totalInputTokens")
    total_output_tokens: int = Field(..., serialization_alias="totalOutputTokens")
    total_cache_creation_tokens: int
    total_cache_read_tokens: int
    total_generation_cost_usd: float
    total_research_tools: int
    total_research_cost_usd: float
    total_cost_usd: float
    cost_per_post: Optional[float] = None

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class RunCostBreakdown(BaseModel):
    """Detailed cost breakdown for a single run"""

    run_id: str = Field(..., serialization_alias="runId")
    project_id: str = Field(..., serialization_alias="projectId")
    status: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    total_input_tokens: int = Field(..., serialization_alias="totalInputTokens")
    total_output_tokens: int = Field(..., serialization_alias="totalOutputTokens")
    total_cache_creation_tokens: int
    total_cache_read_tokens: int
    total_cost_usd: float
    estimated_cost_usd: Optional[float] = None
    total_posts: int = Field(..., serialization_alias="totalPosts")
    posts_with_token_data: int
    avg_cost_per_post: Optional[float] = None
    cache_savings_usd: Optional[float] = Field(
        None, description="Estimated cost savings from prompt caching"
    )

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class CostTrend(BaseModel):
    """Daily cost data point for trends"""

    date: str  # ISO date string (YYYY-MM-DD)
    cost_usd: float


class UserCostSummary(BaseModel):
    """Cost summary across all user's projects"""

    user_id: str = Field(..., serialization_alias="userId")
    period_days: int = Field(description="Number of days analyzed")
    total_projects: int
    total_runs: int = Field(..., serialization_alias="totalRuns")
    total_input_tokens: int = Field(..., serialization_alias="totalInputTokens")
    total_output_tokens: int = Field(..., serialization_alias="totalOutputTokens")
    total_generation_cost_usd: float
    total_research_tools: int
    total_research_cost_usd: float
    total_cost_usd: float
    top_projects: List[Dict[str, Any]] = Field(
        description="Top 5 most expensive projects",
        default_factory=list,
    )
    cost_trend: List[CostTrend] = Field(
        description="Daily cost trend (last 7 days)",
        default_factory=list,
    )

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class ResearchCostSummary(BaseModel):
    """Research tool cost summary for a client"""

    client_id: str = Field(..., serialization_alias="clientId")
    client_name: str = Field(..., serialization_alias="clientName")
    total_research_tools: int = Field(description="Total number of research tools executed")
    total_business_price_usd: float = Field(
        description="Total business model price ($300-600 per tool)"
    )
    total_actual_cost_usd: float = Field(description="Total actual API cost")
    price_difference_usd: float = Field(description="Business price - actual cost (profit margin)")
    tools_breakdown: List[Dict[str, Any]] = Field(
        description="Per-tool cost breakdown",
        default_factory=list,
    )

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)
